File: go-sdk-demo/usermgr/usermgr.go
package usermgr

import (
	"errors"
	"fmt"
	"sort"
	"strings"
	"sync"
)

// ErrNotFound is returned by [Client.Get] when no user matches the given ID.
var ErrNotFound = errors.New("usermgr: user not found")

// ValidationError describes one or more invalid fields on the input.
//
// Callers can match this with [errors.As] to surface field-level details to
// end users.
type ValidationError struct {
	Field   string
	Message string
}

// Error implements the [error] interface.
func (e *ValidationError) Error() string {
	return fmt.Sprintf("usermgr: invalid %s: %s", e.Field, e.Message)
}

// User is the canonical user record returned by the client.
type User struct {
	ID    int64
	Name  string
	Age   int
	Email string
}

// CreateParams carries the fields needed to register a new user.
//
// All fields are required. See [Client.Create] for the validation rules.
type CreateParams struct {
	Name  string
	Age   int
	Email string
}

// ListOptions filters and paginates [Client.List] results.
//
// A zero-value ListOptions returns all users in insertion order.
type ListOptions struct {
	// Keyword filters users whose Name contains this substring (case-insensitive).
	Keyword string
	// Page is 1-based. Zero is treated as 1.
	Page int
	// Size caps results per page. Zero is treated as 20. Maximum is 100.
	Size int
}

// Client is a thread-safe handle to an in-memory user store.
//
// The zero value is not usable; obtain a Client via [New].
type Client struct {
	mu     sync.RWMutex
	nextID int64
	users  map[int64]User
}

// New returns a ready-to-use [Client] with an empty store.
func New() *Client {
	return &Client{users: make(map[int64]User)}
}

// Create validates p and inserts a new user, returning its generated ID.
//
// Validation rules:
//   - Name must be 1–31 characters
//   - Age must be in [0, 200)
//   - Email must contain "@"
//
// On any validation failure Create returns a [*ValidationError] describing the
// first offending field.
func (c *Client) Create(p CreateParams) (int64, error) {
	switch {
	case len(p.Name) == 0 || len(p.Name) > 31:
		return 0, &ValidationError{Field: "Name", Message: "must be 1-31 chars"}
	case p.Age < 0 || p.Age >= 200:
		return 0, &ValidationError{Field: "Age", Message: "must be in [0, 200)"}
	case !strings.Contains(p.Email, "@"):
		return 0, &ValidationError{Field: "Email", Message: "must contain @"}
	}

	c.mu.Lock()
	defer c.mu.Unlock()
	c.nextID++
	id := c.nextID
	c.users[id] = User{ID: id, Name: p.Name, Age: p.Age, Email: p.Email}
	return id, nil
}

// Get returns the user with the given ID.
//
// It returns [ErrNotFound] (wrappable with [errors.Is]) if no such user exists.
func (c *Client) Get(id int64) (User, error) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	u, ok := c.users[id]
	if !ok {
		return User{}, ErrNotFound
	}
	return u, nil
}

// List returns users matching opts plus the total count *before* pagination.
//
// Results are sorted by ID ascending for deterministic output.
func (c *Client) List(opts ListOptions) (users []User, total int) {
	if opts.Page <= 0 {
		opts.Page = 1
	}
	if opts.Size <= 0 {
		opts.Size = 20
	}
	if opts.Size > 100 {
		opts.Size = 100
	}

	c.mu.RLock()
	defer c.mu.RUnlock()

	all := make([]User, 0, len(c.users))
	kw := strings.ToLower(opts.Keyword)
	for _, u := range c.users {
		if kw != "" && !strings.Contains(strings.ToLower(u.Name), kw) {
			continue
		}
		all = append(all, u)
	}
	sort.Slice(all, func(i, j int) bool { return all[i].ID < all[j].ID })

	total = len(all)
	start := (opts.Page - 1) * opts.Size
	if start >= total {
		return nil, total
	}
	end := start + opts.Size
	if end > total {
		end = total
	}
	return all[start:end], total
}

File: ts-sdk-demo/src/usermgr.ts
/**
 * A tiny in-memory user manager — the TypeScript counterpart of
 * `go-sdk-demo/usermgr` and `python-sdk-demo/usermgr`.
 *
 * ## Quick start
 *
 * ```ts
 * import { Client } from 'usermgr'
 *
 * const c = new Client()
 * const id = c.create({ name: 'alice', age: 20, email: 'alice@example.com' })
 * console.log(c.get(id).name)  // 'alice'
 * ```
 *
 * ## Error handling
 *
 * Missing lookups throw {@link NotFoundError}; invalid input throws
 * {@link ValidationError} carrying the offending `field`.
 *
 * @module
 */

/** Raised by {@link Client.get} when no user matches the given ID. */
export class NotFoundError extends Error {
  constructor(id: number) {
    super(`usermgr: user ${id} not found`)
    this.name = 'NotFoundError'
  }
}

/** Raised when input fields fail validation. */
export class ValidationError extends Error {
  /** The offending field, e.g. `"name"`. */
  readonly field: string
  constructor(field: string, message: string) {
    super(`usermgr: invalid ${field}: ${message}`)
    this.name = 'ValidationError'
    this.field = field
  }
}

/** The canonical user record. */
export interface User {
  id: number
  name: string
  age: number
  email: string
}

/**
 * Parameters required to register a new user.
 *
 * Validation rules — see {@link Client.create}:
 * - `name`: 1–31 characters
 * - `age`: integer in `[0, 200)`
 * - `email`: must contain `@`
 */
export interface CreateParams {
  name: string
  age: number
  email: string
}

/**
 * Filter and pagination options for {@link Client.list}.
 *
 * Omitted fields fall back to sensible defaults: no filter, page 1,
 * size 20 (capped at 100).
 */
export interface ListOptions {
  /** Case-insensitive substring filter on `name`. */
  keyword?: string
  /** 1-based page number. */
  page?: number
  /** Items per page (clamped to `[1, 100]`). */
  size?: number
}

/**
 * A thread-safe-ish in-memory user store.
 *
 * (Node's single-threaded event loop makes locking unnecessary; this class
 * mirrors the Go/Python siblings for parity.)
 */
export class Client {
  private nextId = 0
  private users = new Map<number, User>()

  /**
   * Validate `params` and insert a new user.
   *
   * @returns The generated user ID.
   * @throws {ValidationError} If any field is invalid.
   */
  create(params: CreateParams): number {
    if (params.name.length < 1 || params.name.length > 31) {
      throw new ValidationError('name', 'must be 1-31 chars')
    }
    if (!Number.isInteger(params.age) || params.age < 0 || params.age >= 200) {
      throw new ValidationError('age', 'must be in [0, 200)')
    }
    if (!params.email.includes('@')) {
      throw new ValidationError('email', 'must contain @')
    }

    this.nextId += 1
    const id = this.nextId
    this.users.set(id, { id, ...params })
    return id
  }

  /**
   * Return the user with the given ID.
   *
   * @throws {NotFoundError} If no such user exists.
   */
  get(id: number): User {
    const u = this.users.get(id)
    if (!u) throw new NotFoundError(id)
    return u
  }

  /**
   * Return users matching `opts` plus the pre-pagination total.
   *
   * Results are sorted by `id` ascending for deterministic output.
   */
  list(opts: ListOptions = {}): { users: User[]; total: number } {
    const page = opts.page && opts.page > 0 ? opts.page : 1
    const sizeRaw = opts.size && opts.size > 0 ? opts.size : 20
    const size = Math.min(sizeRaw, 100)
    const kw = (opts.keyword ?? '').toLowerCase()

    const all = [...this.users.values()]
      .filter((u) => kw === '' || u.name.toLowerCase().includes(kw))
      .sort((a, b) => a.id - b.id)

    const start = (page - 1) * size
    return { users: all.slice(start, start + size), total: all.length }
  }
}
